autofaiss/external/optimize.py

```diff
@@ -483,7 +483,7 @@ def optimize_and_measure_indices(
     ]
 
     def _read_one_index(index_file_path: str):
-        with fsspec.open(index_file_path, "rb").open() as f:
```

Comment:
Question: out of curiosity, does `fsspec.open` work ? or it is just fs.open() is faster. I asked this because we have several other places using `fsspec.open`.

Reply:
good question...
in fact I just checked and this won't solve my issue
the problem is that 
```python
    indices_file_paths = [
        prefix.rstrip("/") + "/" + index_file_path.lstrip("/") for index_file_path in sorted(indices_file_paths)
    ]
```
is incorrect

it produces things like `s3:/laion-us-east-1/mytest/my_tmp_folder5/index_447` (one missing /)

I'm fixing that

BTW, does this N indice feature work end to end on your side ?
I'm having issues with the last steps (merging and optimizing) hence this PR

Reply:
I think we need https://github.com/criteo/autofaiss/issues/103 to iterate faster

Reply:
but yes in general fsspec.open works, but I usually avoids it as it requires prefixed paths and instead use fsspec.core.url_to_fs when fs.open so I can move around only relative paths

Reply:
pushed a new commit with a fix

Reply:
> it produces things like s3:/laion-us-east-1/mytest/my_tmp_folder5/index_447 (one missing /)

I don't quite understand what it should be for paths ?

> BTW, does this N indice feature work end to end on your side ?

Good question, on our side, 
1) for indexing building, we have tried N-indices, it didn't work last week due to out of disk error, today we tried the version 2.10.3, it didn't work until the end, I am still trying to find the error log (which is truncated, you know well :) ) So it is probably caused by the line you mentioned.
2)  for the indexing searching part, searching from N-indices shows very promising results, no Korean or Russian language is dominant in the result.

Reply:
> for the indexing searching part, searching from N-indices shows very promising results, no Korean or Russian language is dominant in the result.

nice!
on my side, I tried various level of PQ for a 5B index : PQ24, PQ64 and PQ128, of course with increasing size of indices (PQ 128 is 800GB)
I found that PQ24 is not working super well (best similarities are around 0.1), PQ64 is ok (best sim at 0.2) and PQ128 is good (best sim at 0.28)
for 400m points in the past I used PQ256, but it was a bit overkill

I might implement the recall feature in order to measure precisely what's impacting things. Increasing the PQ value works but it's kind of the brute force solution. It is possible some other params could be tweaked to get better recall (like number of clusters, number of points used for training, ...)

Reply:
manipulating a 800GB index is possible but not the most convenient (even if kept as 20 indices), so I don't know if I would recommend it to you, you need a lot of disk space

Reply:
Good, right now what it is suggested by autofaiss is `PQ48` with max_index_memory_usage 250GB, current_memory_available 70GB and 5 indices_to_keep.

I just know that our last run failed due to OOM at the optimising step. hmm...

Reply:
> I see what you mean by missing one / now, it should be something like s3:// right ?
yes

I just removed this whole prefixing logic now so it should be better

Reply:
btw I'm @rom1504 at https://discord.gg/eq3cAMZtCC ; might be easier to discuss some details in instant messaging
(even though I believe all these changes/fixes are almost done now, maybe it's the last one needed)

Reply:
> current_memory_available 70GB and 5 indices_to_keep.

yeah, let's fix the optimization by moving it to distributed, but maybe still increase to 10 indices, 70GB is a bit close to the limit